lib/guardian.ml

```diff
@@ -459,108 +418,64 @@ module Make (A : RoleSig) (T : RoleSig) = struct
+      let expand =
+        Lwt_list.map_s (fun effect ->
+          match effect with
+          | action, `Target (x : Uuid.Target.t) ->
+            actor_of x
+            |> Actor.find_roles ?ctx
+            (* search for roles if target would be an actor, default: empty *)
+            >|= CCResult.get_or ~default:ActorRoleSet.empty
+            >|= ActorRoleSet.elements
+            >>= Lwt_list.map_s (fun role ->
+                  expand_effects
+                    ?ctx
+                    [ action, `TargetEntity (role |> A.show |> T.of_string) ])
+            >|= flatten
+            |> Lwt_result.map
+                 (CCList.fold_left (flip add) (singleton effect) %> elements)
+          | x -> Lwt.return_ok [ x ])
       in
-      CCList.fold_left
-        (fun x acc ->
-          match acc, x with
-          | Error _, _ -> acc
-          | Ok _, Error _ -> x
-          | Ok acc', Ok x' -> Ok (x' @ acc'))
-        (Ok [])
-        lsls
-      |> Lwt.return
+      effects
+      |> expand
+      >|= flatten
+      |> Lwt_result.map_error
+           (Format.asprintf
+              "Failed to expand the effects of the target. Error message: %s")
     ;;
 
     (** [collect_rules e] Query the database for a list of rules pertaining to
         the effects [e]. *)
     let collect_rules ?ctx (effects : Authorizer.effect list) =
+      let open CCFun in
+      let flatten = CCResult.flatten_l %> CCResult.map CCList.flatten in
       Lwt_result.map_error
         (Format.asprintf
            "Failed to collect rules for effects list %s. Error message: %s"
            ([%show: Authorizer.effect list] effects))
       @@
-      let open Lwt_result.Syntax in
-      let* effects = expand_effects ?ctx effects in
-      let%lwt results =
-        Lwt_list.map_s
-          (fun (action, target) ->
-            let* rules = Actor.find_rules ?ctx target in
-            CCList.filter
-              (fun (_actor, action', _target) ->
-                action = action' || action' = `Manage)
-              rules
-            |> Lwt_result.return)
-          effects
-      in
-      let* rules =
-        CCList.fold_left
-          (fun acc (x : (auth_rule list, string) result) ->
-            let open CCResult in
-            both acc x >|= CCFun.uncurry CCList.append)
-          (Ok [])
-          results
-        |> Lwt_result.lift
-      in
-      CCList.fold_right
-        Authorizer.Auth_rule_set.add
-        rules
-        Authorizer.Auth_rule_set.empty
-      |> Authorizer.Auth_rule_set.elements
-      |> Lwt_result.return
+      let open Lwt_result.Infix in
+      let open Authorizer.Auth_rule_set in
+      effects
+      |> expand_effects ?ctx
+      >>= Lwt_list.map_s (fun (action, target) ->
+            Actor.find_rules ?ctx target
+            >|= CCList.filter (fun (_, rule_action, _) ->
+                  action = rule_action || rule_action = `Manage))
+          %> Lwt.map flatten
```

Comment:
Wow, this is much shorter than that `both acc x >|= CCFun.uncurry CCList.append` in a fold :+1: